allennlp/models/encoder_decoders/wikitables_semantic_parser.py

```diff
@@ -269,46 +318,96 @@ def __init__(self,
                  batch_indices: List[int],
                  action_history: List[List[int]],
                  score: Variable,
+                 non_terminal_stack: List[List[str]],
                  hidden_state: List[Variable],
                  memory_cell: List[Variable],
                  attended_question: List[Variable],
                  encoder_outputs: Variable,
                  encoder_output_mask: Variable,
+                 global_type_productions: Dict[str, List[int]],
                  end_index: int) -> None:
         super(WikiTablesDecoderState, self).__init__(batch_indices, action_history, score)
+        self.non_terminal_stack = non_terminal_stack
         self.hidden_state = hidden_state
         self.memory_cell = memory_cell
         self.attended_question = attended_question
         self.encoder_outputs = encoder_outputs
         self.encoder_output_mask = encoder_output_mask
+        self.global_type_productions = global_type_productions
         self.end_index = end_index
 
-    def get_valid_actions(self, num_actions: int) -> List[List[int]]:
+    def get_valid_actions(self) -> List[List[int]]:
         """
         Returns a list of valid actions for each element of the group.
         """
-        # TODO(mattg): we're temporarily making all actions allowed at every state, to implement
-        # constrained decoding in pieces.  The `num_actions` argument here is going to be removed
-        # in the next step, where we actually use grammar constraints.
-        return [list(range(num_actions))] * len(self.batch_indices)
+        # TODO(matt): this is going to need to look at more than just the global type productions
+        # eventually - we'll need instance- and state-specific additions.  And we'll probably need
+        # to return strings, instead of integers, because instance-specific production rules won't
+        # be in a global vocabulary.
+        return [self.global_type_productions[stack[-1]] for stack in self.non_terminal_stack]
+
+    @staticmethod
+    def update_non_terminal_stack(stack: List[str], action: str) -> List[str]:
+        """
+        Given a single non-terminal stack (`not` a grouped one), and an action that was taken,
+        update the non-terminal stack.  This involves popping the non-terminal that was expanded
+        off of the stack, then pushing on any non-terminals in the production rule back on the
+        stack.
+        """
+        if ' -> ' in action:
+            non_terminal, production_string = action.split(' -> ')
```

Comment:
I think you'll have to keep track of the type of the variable when the production string contains a lambda here. Something to keep in mind for the future.

Reply:
Added a TODO.